### PR TITLE
Fast Confirmation Rule Utilities part 1

### PR DESCRIPTION
**Pull request by zilm13**
## PR Description
Started slicing #10948 
First part
100% isolated utilities just following the spec, see:
https://github.com/ethereum/consensus-specs/blob/master/specs/phase0/fast-confirmation.md
https://github.com/ethereum/consensus-specs/blob/master/specs/bellatrix/fast-confirmation.md
https://github.com/ethereum/consensus-specs/blob/master/specs/gloas/fast-confirmation.md

UPDATED: dropped Calculator to reduce size. It will go in separate PR

## Documentation

- [x] I thought about documentation and added the `doc-change-required` label to this PR if updates are required.

## Changelog

- [x] I thought about adding a changelog entry, and added one if I deemed necessary.

---

> [!NOTE]
> **Medium Risk**
> Touches fork-choice store vote access and documents Byzantine threshold as a hardcoded constant; behavior is mostly additive and not yet integrated into head/confirmation decisions.
> 
> **Overview**
> Adds the first slice of **Fast Confirmation** support: spec-aligned helpers and data types, without wiring `on_fast_confirmation` into live fork choice yet.
> 
> Introduces an immutable **`FastConfirmationStore`** record (confirmed root, epoch observed justified checkpoints, slot heads) with factory/`with*` updates, plus **`FastConfirmationStates`** to bundle the beacon states the rule will read per slot. **`FastConfirmationRuleUtil`** implements early spec pieces: committee weight estimation across slots/epochs, safety adjustment, greatest unrealized justified checkpoint from fork-choice block metadata, and **`update_fast_confirmation_variables`** (slot head rotation and epoch-boundary checkpoint rotation).
> 
> Fork-choice **`ReadOnlyStore`** gains **`getVoteSnapshot()`** backed by a new immutable **`VoteSnapshot`** (copy of votes up to the highest voted index). **`Store`** and test fixtures implement it; **`StoreTransaction`** deliberately returns only **committed** votes so uncommitted vote updates are invisible to the rule. **`specrefs`** entries are updated to point at these implementations for several fast-confirmation symbols.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit cda31827987157be6f3ea6982ff9d2764712a2ce. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

File: ethereum/statetransition/src/main/java/tech/pegasys/teku/statetransition/forkchoice/fastconfirmation/FastConfirmationCalculator.java [hunk 1]
```diff
@@ -0,0 +1,715 @@
...
+      if (vote.isEquivocating()) {
+        continue;
+      }
+      final Bytes32 votedRoot = vote.getNextRoot();
+      // A zero root means the validator is not in store.latest_messages.
+      if (!votedRoot.isZero() && isAncestor(votedRoot, nodeRoot)) {
+        score = score.plus(validator.getEffectiveBalance());
+      }
+    }
+    return score;
```
[the remark this answers is not in the document]
Comment (zilm13) [reply ethereum/statetransition/src/main/java/tech/pegasys/teku/statetransition/forkchoice/fastconfirmation/FastConfirmationCalculator.java]: `get_supported_node` is not used in Fast Confirmation Rule. There are no changes like mentioned in spec https://github.com/ethereum/consensus-specs/blob/master/specs/gloas/fast-confirmation.md

[the remark this answers is not in the document]
Comment (zilm13) [reply ethereum/statetransition/src/main/java/tech/pegasys/teku/statetransition/forkchoice/fastconfirmation/FastConfirmationCalculator.java]: Ok, I see it's `FastConfirmationCalculator.getAttestationScore` implementing `get_attestation_score` separately from `ForkChoice`. I will recheck it in next pr which will include `FastConfirmationCalculator`. It's no more a part of this PR

**Comment by rolfyone** [comment]
ran a review and got a couple of things, but not sure if they're addressed in other PRs, so will just share it here

---
  Overview: Solid first slice of the FCR implementation — the utility functions are spec-faithful and the VoteSnapshot thread-safety model is correct. A few things worth addressing before merge:

  Medium — updateFastConfirmationVariables uses raw 7-arg constructors
  The method creates up to 3 FastConfirmationStore records using the positional constructor. Argument-order mistakes here would be silent. The record already has a withConfirmedRoot() helper; adding with* helpers for the other fields would make this
  much safer and easier to read.

  Low — CONFIRMATION_BYZANTINE_THRESHOLD hardcoded
  The spec exposes this as a config variable. The comment explains the rationale (both presets are 25), but if a future testnet ever deviates, the FCR will silently use the wrong value. Worth noting in the comment that it would need manual updating
  in that case, or tracking a follow-up to move it into SpecConfig.

  Low — StoreTransaction.getVoteSnapshot() silently excludes pending votes
  It delegates directly to the underlying store, so uncommitted vote changes in the transaction are not included. This is probably correct (FCR should see committed vote state), but a one-line comment would save a future reader from wondering
  whether this is intentional.

  Low — missing edge-case test for estimateCommitteeWeightBetweenSlots
  The existing cross-epoch test uses 2 slots from each epoch (6..9). A case covering nearly a full epoch in the start epoch (e.g. 1..9 — 7 of 8 slots in epoch 0 plus 2 from epoch 1) would exercise the pro-rata formula at extremes where
  integer-division truncation is largest.

**Review by rolfyone — DISMISSED** [review]

**Comment by zilm13** [comment]
@rolfyone I've addressed feedback, see https://github.com/Consensys/teku/pull/10984/commits/97284e6570de87695149e94508d9785aead51d4d
I've kept `CONFIRMATION_BYZANTINE_THRESHOLD` hardcoded, just extended the comment to make it clear. Otherwise we should put it to all 10 configs because you are able to run FCR in any phase0 network. I could do this but I prefer to be less invasive at the current stage. What do you think?

**Review by mehdi-aouadi — APPROVED** [review]
LGTM
